Objective-C/CBLDatabase.mm

```diff
@@ -360,58 +360,89 @@ - (BOOL) inBatch: (NSError**)outError usingBlock: (void (NS_NOESCAPE ^)())block
 #pragma mark - DATABASE MAINTENANCE
 
 - (BOOL) close: (NSError**)outError {
+    NSArray *activeReplicators, *activeLiveQueries = nil;
+    
     CBL_LOCK(self) {
-        if (_c4db == nullptr)
+        if ([self isClosed])
             return YES;
         
         CBLLogInfo(Database, @"Closing %@ at path %@", self, self.path);
-    
-        if (_activeReplications.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageCloseDBFailedReplications, outError);
-        }
         
-        if (_liveQueries.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageCloseDBFailedQueryListeners, outError);
+        if (!_isClosing) {
+            _isClosing = YES;
+            
+            if (!_closeCondition)
+                _closeCondition = [[NSCondition alloc] init];
+            
+            activeReplicators = [_activeReplicators allObjects];
+            
+            activeLiveQueries = [_activeLiveQueries allObjects];
         }
-        
-        [self cancelDocExpiryTimer];
+    }
     
-        C4Error err;
-        if (!c4db_close(_c4db, &err))
-            return convertError(err, outError);
+    // Stop active replicators:
+    for (CBLReplicator* r in activeReplicators) {
+        [r stop];
+    }
     
-        [self freeC4Observer];
-        [self freeC4DB];
+    // Stop active live queries:
+    for (CBLLiveQuery* q in activeLiveQueries) {
+        [q stop];
+    }
     
-        return YES;
+    // Wait for all active replicators and live queries to stop:
+    [_closeCondition lock];
+    while (![self isReadyToClose]) {
+        [_closeCondition wait];
     }
-}
-
-- (BOOL) delete: (NSError**)outError {
+    [_closeCondition unlock];
+    
     CBL_LOCK(self) {
-        [self mustBeOpen];
-        
-        if (_activeReplications.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageDeleteDBFailedReplications, outError);
-        }
-        
-        if (_liveQueries.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageDeleteDBFailedQueryListeners, outError);
-        }
+        if ([self isClosed])
+            return YES;
         
+        // Cancel doc expiry timer:
         [self cancelDocExpiryTimer];
         
+        // Free C4Observer:
+        [self freeC4Observer];
+        
+        // Close database:
+        BOOL success = YES;
         C4Error err;
-        if (!c4db_delete(_c4db, &err))
-            return convertError(err, outError);
+        if (!c4db_close(_c4db, &err))
+            success = convertError(err, outError);
         
-        [self freeC4Observer];
-        [self freeC4DB];
+        // Release database:
+        if (success)
+            [self freeC4DB];
         
-        return YES;
+        // Reset closing flag:
+        _isClosing = NO;
+        
+        return success;
     }
 }
 
+- (BOOL) isReadyToClose {
+    CBL_LOCK(self) {
```

Comment:
wondering whether we need this lock, since the condition lock should already be locking this predicate check.

Reply:
condition lock doesn't guard _activeReplicators and _activeLiveQueries from accessing by multiple threads. It is required to lock here.

Reply:
true!